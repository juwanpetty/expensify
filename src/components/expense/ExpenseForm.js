import React, { Component } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

class ExpenseForm extends Component {
  constructor(props) {
    super();

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      cretaedAt: props.expense ? moment(props.expense.cretaedAt) : moment(),
      calendarFocused: false,
      formError: null
    };
  }

  onDescriptionChange = event => {
    const description = event.target.value;
    this.setState({
      description
    });
  };

  onNoteChange = event => {
    const note = event.target.value;
    this.setState({
      note
    });
  };

  onAmountChange = event => {
    const amount = event.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState({
        amount
      });
    }
  };

  onDateChange = cretaedAt => {
    if (cretaedAt) {
      this.setState({
        cretaedAt
      });
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState({ calendarFocused: focused });
  };

  onSubmit = event => {
    event.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState({
        formError: "Please provide description and amount"
      });
    } else {
      this.setState({
        formError: null
      });

      this.props.onSubmit({
        description: this.state.description,
        note: this.state.note,
        amount: parseFloat(this.state.amount, 10) * 100,
        cretaedAt: this.state.cretaedAt.valueOf()
      });
    }
  };

  render() {
    const errorMarkup = this.state.formError ? (
      <p>{this.state.formError}</p>
    ) : null;

    return (
      <div>
        {errorMarkup}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            name=""
            id=""
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="number"
            name=""
            id=""
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.cretaedAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a not for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
