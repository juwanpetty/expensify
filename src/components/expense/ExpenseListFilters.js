import React, { Component } from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import {
  setTextFilter,
  setStartDate,
  setEndDate,
  sortByAmount,
  sortByDate
} from "../../store/actions/filtersAction";

class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  };

  onChangeInput = event => {
    this.props.dispatch(setTextFilter(event.target.value));
    console.log(event.target.value);
  };

  onChangeSelect = event => {
    if (event.target.value === "date") {
      this.props.dispatch(sortByDate());
    } else {
      this.props.dispatch(sortByAmount());
    }

    console.log(event.target.value);
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };

  onFocusChange = calendarFocused => {
    this.setState({
      calendarFocused
    });
  };

  render() {
    return (
      <div>
        <input
          type="text"
          onChange={event => this.onChangeInput(event)}
          value={this.props.filters.text}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={event => this.onChangeSelect(event)}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={() => this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseListFilters);
