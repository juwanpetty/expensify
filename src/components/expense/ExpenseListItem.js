import React from "react";
import { connect } from "react-redux";
import { removeExpense } from "../../store/actions/expensesAction";

const ExpenseListItem = ({ description, amount, createdAt, id, dispatch }) => {
  const removeItem = () => {
    console.log("removing item...", { id, description, amount, createdAt });
    dispatch(removeExpense({ id }));
  };

  return (
    <div>
      <h3>{description}</h3>
      <p>
        {amount} - {createdAt}
      </p>
      <button onClick={() => removeItem()}>Remove</button>
    </div>
  );
};

export default connect()(ExpenseListItem);
