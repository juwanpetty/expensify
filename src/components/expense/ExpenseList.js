import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "../expense/ExpenseListItem";
import selectExpenses from "../../store/selectors/expensesSelector";

export const ExpenseList = ({ expenses }) => {
  return (
    <div>
      {expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        expenses.map(expense => (
          <ExpenseListItem key={expense.id} {...expense} />
        ))
      )}
    </div>
  );
};

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
