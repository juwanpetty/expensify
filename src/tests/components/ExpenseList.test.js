import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/expense/ExpenseList";
import expenses from "../fixtures/expenses";

it("should render expense list with expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

it("should render expense list with empty message", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
