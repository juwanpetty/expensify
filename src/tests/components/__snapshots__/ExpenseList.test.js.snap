// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list with empty message 1`] = `ShallowWrapper {}`;

exports[`should render expense list with expenses 1`] = `ShallowWrapper {}`;
